subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorDetector.java

```diff
@@ -0,0 +1,128 @@
+import org.gradle.util.DeprecationLogger;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Inspects a classpath to find annotation processors contained in it.
+ */
+public class AnnotationProcessorDetector {
+    private final FileContentCache<List<AnnotationProcessorDeclaration>> cache;
+
+    public AnnotationProcessorDetector(FileContentCacheFactory cacheFactory) {
+        cache = cacheFactory.newCache("annotation-processors", 20000, new ProcessorServiceLocator(), new ListSerializer<AnnotationProcessorDeclaration>(new AnnotationProcessorDeclarationSerializer()));
+    }
+
+    public List<AnnotationProcessorDeclaration> detectProcessors(Iterable<File> processorPath) {
+        List<AnnotationProcessorDeclaration> processors = Lists.newArrayList();
+        for (File jarOrClassesDir : processorPath) {
+            processors.addAll(cache.get(jarOrClassesDir));
+        }
+        return processors;
+    }
+
+    private static class ProcessorServiceLocator implements FileContentCacheFactory.Calculator<List<AnnotationProcessorDeclaration>> {
+        @Override
+        public List<AnnotationProcessorDeclaration> calculate(File file, FileType fileType) {
+            if (fileType == FileType.Directory) {
+                return detectProcessorsInClassesDir(file);
+            }
+            if (fileType == FileType.RegularFile && FileUtils.hasExtensionIgnoresCase(file.getName(), ".jar")) {
+                return detectProcessorsInJar(file);
+            }
+            return Collections.emptyList();
+        }
+
+        private List<AnnotationProcessorDeclaration> detectProcessorsInClassesDir(File classesDir) {
+            File processorDeclaration = new File(classesDir, "META-INF/services/javax.annotation.processing.Processor");
+            if (!processorDeclaration.isFile()) {
+                return Collections.emptyList();
+            }
+            List<String> processorNames = readLines(processorDeclaration);
+            return toProcessorDeclarations(processorNames);
+        }
+
+        private List<String> readLines(File file) {
+            try {
+                return Files.readLines(file, Charsets.UTF_8);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+
+        private List<AnnotationProcessorDeclaration> detectProcessorsInJar(File jar) {
+            try {
+                ZipFile zipFile = new ZipFile(jar);
+                try {
+                    return detectProcessorsInZipFile(zipFile);
+                } finally {
+                    zipFile.close();
+                }
+            } catch (IOException e) {
+                DeprecationLogger.nagUserWith("Malformed jar [" + jar.getName() + "] found on compile classpath. Gradle 5.0 will no longer allow malformed jars on the compile classpath.");
```

Comment:
I think we need to revisit this. We have a ticket reporting that this can be intentional (placing a corrupted jar on classpath for tests). Or a way to disable.

Reply:
Can you point me to the ticket? I think a good way would be to log a message instead that says "We'll assume this jar contains no processors".

Reply:
Actually we miss a ticket. I thought the user had created one, but it seems not: https://twitter.com/CedricChampeau/status/943149158805966849

Reply:
Can you open one and assign it to me?

Reply:
Here you go: https://github.com/gradle/gradle/issues/4218